src/components/slide-carousel/SlideCarousel.js

```diff
@@ -35,24 +40,38 @@ class SlideCarousel extends React.Component<Props, State> {
         };
     }
 
+    isControlled() {
+        const { currentIndex } = this.props;
+        return currentIndex !== undefined;
+    }
+
     /*
      * If the selected index in the state has somehow gotten set to an
      * out of bounds value (either because we were passed a bad value,
      * or the number of children has reduced), compute a new selected
      * index which is a floored value between 0 <= index < num children
      */
     getBoundedSelectedIndex() {
-        const { children } = this.props;
+        const { children, currentIndex } = this.props;
         const { selectedIndex } = this.state;
 
+        const indexToCheck = this.isControlled() ? currentIndex : selectedIndex;
+
         const lastChildIndex = Math.max(React.Children.count(children) - 1, 0);
-        const boundedSelectedIndex = Math.max(selectedIndex || 0, 0);
+        const boundedSelectedIndex = Math.max(indexToCheck || 0, 0);
 
         return boundedSelectedIndex > lastChildIndex ? lastChildIndex : Math.floor(boundedSelectedIndex);
     }
 
     setSelectedIndex = (index: number) => {
-        this.setState({ selectedIndex: index });
+        const { onSlideChange } = this.props;
+        if (!this.isControlled()) {
+            // Only update local state if we're using this as an uncontrolled component
+            this.setState({ selectedIndex: index });
+        }
+        if (onSlideChange) {
```

Comment:
Should this be part of the Primitive instead? Technically the primitive is sorta the controlled version of this component. Maybe a bit of function re-shuffle between this file and the primitive might just make the Primitive what you want?